[DBZ-1408] Refactor SourceInfo to extend AbstractSourceInfo

bingqinzhou opened the pull request:
https://issues.jboss.org/projects/DBZ/issues/DBZ-1408

This PR includes the following major changes:
- Refactor CassandraConnectorConfig to extend CommonConnectorConfig.
- Refactor SourceInfo to extend AbstractSourceInfo.

gunnarmorling commented:
Hey @criccomini and @jgao54, any feedback from your side on this one?

@bingqinzhou, could you rebase it to the latest master, as there's a conflict. Thanks!

gunnarmorling commented on debezium-connector-cassandra/src/main/java/io/debezium/connector/cassandra/CassandraConnectorConfig.java:
```diff
@@ -373,70 +357,72 @@ public CommitLogTransfer getCommitLogTransfer() {
+     * during compaction if they haven't been consumed by the consumer yet. In other words, do NOT
+     * set this to true if you have kafka compaction turned on.
+     */
     public boolean tombstonesOnDelete() {
```
That's already handled by `CommonConnectorConfig` (see `isEmitTombstoneOnDelete()` and related code).

bingqinzhou replied:
> isEmitTombstoneOnDelete

We'll need to rewrite one here because the default value of `TOMBSTONES_ON_DELETE` in other Debezium Connectors is `true`, but in Cassandra Connector is `false`.

gunnarmorling replied:
Ah, would be nice to align the default then. Is this still something you wanted to do?

bingqinzhou replied:
Make sense! Updated to align the default to be `true` as other debezium connectors. 
But want to double confirm with @jgao54 if it is ok to set the default value of `TOMBSTONES_ON_DELETE` to be `true` ..?

bingqinzhou replied:
Make sense! Updated to align the default to be `true` as other debezium connectors. 
But want to double confirm with @jgao54 if it is ok to set the default value of `TOMBSTONES_ON_DELETE` to be `true` ..?

bingqinzhou replied:
> Ah, would be nice to align the default then. Is this still something you wanted to do?

Make sense! Updated to align the default to be true as other debezium connectors.
But want to double confirm with @jgao54 if it is ok to set the default value of TOMBSTONES_ON_DELETE to be true ..?

jgao54 replied:
@bingqinzhou  @gunnarmorling 

Assume my understanding of Kafka tombstone and compaction is correct -- such that when a tombstone is detected, it will remove _all_ events in the log for the key _except for the most recent event with the same key_. If so, I'm a bit concerned with setting it to `true` as default, as it may result in unexpected data losses.

The reason is that the value of a CDC event may not contain all the columns, so the most recent event != the most recent row for a given key.

gunnarmorling replied:
Ah, yes. That's a good point. It seems `false` is the saver default then indeed.

gunnarmorling replied:
In other words, the suggested change here is good AFAICS.

gunnarmorling commented on debezium-connector-cassandra/src/main/java/io/debezium/connector/cassandra/CassandraConnectorConfig.java:
```diff
@@ -461,16 +447,31 @@ public Converter getValueConverter() throws CassandraConnectorConfigException {
+
+    @Override
+    protected SourceInfoStructMaker<? extends AbstractSourceInfo> getSourceInfoStructMaker(Version version) {
+        switch (version) {
```
This shouldn't be needed, or do we actually want to maintain two different versions of the source struct? I don't think so, given the connector's "incubating" state. The notion of the "legacy" struct maker exists for the other connectors, as we did some technically non-backwards compatible changes to `source` in 0.10 (for the sake of general consistency and stream-lining, this might btw. relevant for you when upgrading your MySQL connectors) and we wanted to have a transitioning period where the old one still is available, too. This shouldn't be needed for Cassandra.

bingqinzhou replied:
> This shouldn't be needed, or do we actually want to maintain two different versions of the source struct? I don't think so, given the connector's "incubating" state. The notion of the "legacy" struct maker exists for the other connectors, as we did some technically non-backwards compatible changes to `source` in 0.10 (for the sake of general consistency and stream-lining, this might btw. relevant for you when upgrading your MySQL connectors) and we wanted to have a transitioning period where the old one still is available, too. This shouldn't be needed for Cassandra.

Understood, updated accordingly!

gunnarmorling commented on debezium-connector-cassandra/src/main/java/io/debezium/connector/cassandra/LegacyV1CassandraSourceInfoStructMaker.java:
```diff
@@ -0,0 +1,50 @@
+import org.apache.kafka.connect.data.Schema;
+import org.apache.kafka.connect.data.Struct;
+
+public class LegacyV1CassandraSourceInfoStructMaker extends LegacyV1AbstractSourceInfoStructMaker<SourceInfo> {
```
See my comment above; shouldn't be needed.

bingqinzhou replied:
> See my comment above; shouldn't be needed.

Removed LegacyV1CassandraSourceInfoStructMaker accordingly.

gunnarmorling commented on debezium-connector-cassandra/src/main/java/io/debezium/connector/cassandra/SourceInfo.java:
```diff
@@ -108,4 +84,19 @@ public String toString() {
+
+    @Override
+    protected String database() {
+        return "NULL";
```
Could you add a clarifying comment; is there no notion of "database" in Cassandra?

bingqinzhou replied:
> Could you add a clarifying comment; is there no notion of "database" in Cassandra?

Got it, added clarifying comment accordingly.

gunnarmorling commented on debezium-connector-cassandra/src/main/java/io/debezium/connector/cassandra/SourceInfo.java:
```diff
@@ -28,49 +27,26 @@
+    public OffsetPosition offsetPosition;
+    public KeyspaceTable keyspaceTable;
+    public boolean snapshot;
+    public long tsMicro;
```
This should be `Instant` better. We try to convert any timestamps to `Instant` right at the system boundary as early as possible and then use that more meaningful type internally. Only where needed, it's "rendered" as long timestamp again.

gunnarmorling replied:
@bingqinzhou could you still address this remark?

bingqinzhou replied:
@gunnarmorling Updated accordingly.

gunnarmorling commented on debezium-connector-cassandra/src/main/java/io/debezium/connector/cassandra/SourceInfo.java:
```diff
@@ -108,4 +84,19 @@ public String toString() {
+
+    @Override
+    protected Instant timestamp() {
+        return Instant.ofEpochSecond(tsMicro/1000);
```
See above, also should be converted keeping the precision.

bingqinzhou replied:
> See above, also should be converted keeping the precision.

Nice catch! Updated to use `Instant.ofEpochMilli(tsMicro)` to keep the precision.

gunnarmorling replied:
But `ofEpochMilli()` will loose precision, no?

bingqinzhou replied:
Oops, my bad. Updated to convert `micro-seconds` to `Instant`.

bingqinzhou replied:
> But `ofEpochMilli()` will loose precision, no?

Oops, my bad. Updated to convert `micro-seconds` to `Instant`.

bingqinzhou commented:
> Hey @criccomini and @jgao54, any feedback from your side on this one?
> 
> @bingqinzhou, could you rebase it to the latest master, as there's a conflict. Thanks!

Sure, rebased and the conflict is resolved now :)

gunnarmorling commented:
> Rebased and the conflict is resolved now :)

Did you push this? Or maybe there's a new conflict (although I don't remember changing anything in Cassandra lately)?

bingqinzhou commented:
> > Rebased and the conflict is resolved now :)
> 
> Did you push this? Or maybe there's a new conflict (although I don't remember changing anything in Cassandra lately)?

@gunnarmorling Hm, are you still seeing any conflicts of this PR..? I resolved the conflicts 12 days ago (https://github.com/debezium/debezium-incubator/pull/109/commits/2ccb2cf14a329f4b9d521b0ff03d389b71c7f6d3) and pushed the change, currently I don't see there's any conflicts with this PR.

jgao54 commented:
@bingqinzhou  sorry for the late response, will review this PR in more details by EOD today.

jgao54 commented:
LGTM!

For data integrity, i think we should have default value for tombstone set to false, given users may not be checking every config when they initially launch the connector and if they have kafka log compaction turned on, they'd experience potential data loss.

bingqinzhou commented:
> LGTM!
> 
> For data integrity, i think we should have default value for tombstone set to false, given users may not be checking every config when they initially launch the connector and if they have kafka log compaction turned on, they'd experience potential data loss.

Make sense, reverted my previous commit and changed the default value of tombstone config to be false :)

bingqinzhou commented:
@gunnarmorling I kept the default value of `tombstoneOnDelete` to be `false` based on @jgao54's explanation about this. I've also made changes according to your previous comments as well as resolved the conflicts. If this looks good to you now, may I get this PR approved and merged please? Otherwise, please let me know if I need to make any further changes. Thank you!

criccomini commented:
Checking in here. @gunnarmorling is this PR OK?

gunnarmorling commented:
Yes, it is from my PoV. Just needs a rebase. @bingqinzhou, once you've done that, feel free to merge it yourself or @criccomini can do it? In addition, an update of the connector docs in the core repo probably is needed to reflect the updated source structure (assuming there's an example of that shown in there).

gunnarmorling commented:
@bingqinzhou, @criccomini, so I went ahead and rebased this and squashed all into one commit. Could either of you double-check on this PR and merge it:

    https://github.com/debezium/debezium-incubator/pull/117

Thanks!

@bingqinzhou, btw. we're preferring rebases over merge commits for the sake of cleaner history. Let me know if you have any trouble with that in case you haven't used rebases before.

bingqinzhou commented:
> @bingqinzhou, @criccomini, so I went ahead and rebased this and squashed all into one commit. Could either of you double-check on this PR and merge it:
> 
> #117
> 
> Thanks!
> 
> @bingqinzhou, btw. we're preferring rebases over merge commits for the sake of cleaner history. Let me know if you have any trouble with that in case you haven't used rebases before.

@gunnarmorling I'm actually not sure how to rebase for a branch in a fork repo. Here's what I'm currently do to update with master:
```
$ git fetch upstream
$ git merge upstream/master
```
Please let me know if there's a better way that I'm supposed to use. Thank you!
